optuna/_callbacks.py

```diff
@@ -55,3 +56,84 @@ def __call__(self, study: "optuna.study.Study", trial: FrozenTrial) -> None:
         n_complete = len(trials)
         if n_complete >= self._n_trials:
             study.stop()
+
+
+@experimental("2.8.0")
+class RetryFailedTrialCallback:
+    """Retry a failed trial up to a maximum number of times.
+
+    When a trial fails, this callback can be used with the :class:`optuna.storage` class to
+    recreate the trial in :obj:`TrialState.WAITING` to queue up the trial to be run again.
+
+    This is helpful in environments where trials may fail due to external conditions, such as
+    being pre-empted by other processes.
+
+    Usage:
+
+        .. testcode::
+
+            import optuna
+            from optuna.storages import RetryFailedTrialCallback
+
+            storage = optuna.storages.RDBStorage(
+                url="sqlite:///:memory:",
+                heartbeat_interval=60,
+                grace_period=120,
+                failed_trial_callback=RetryFailedTrialCallback(max_retry=3),
+            )
+
+            study = optuna.create_study(
+                storage=storage,
+            )
+
+    Args:
+        max_retry:
+            The max number of times a trial can be retried. Must be set to :obj:`None` or an
+            integer. If set to the default value of :obj:`None` will retry indefinitely.
+            If set to an integer, will only retry that many times.
+    """
+
+    def __init__(self, max_retry: Optional[int] = None) -> None:
+        self._max_retry = max_retry
+
+    def __call__(self, study: "optuna.study.Study", trial: FrozenTrial) -> None:
+        system_attrs = {"failed_trial": trial.number}
+
+        # Update the new object with the values in the trial.system_attrs.
+        # By doing this, if this failed try is already a rety, the 'failed_trial' value
+        # will be the first failed trial number.
+        system_attrs.update(trial.system_attrs)
+
+        retries = sum(
+            ("failed_trial", system_attrs["failed_trial"]) in s.system_attrs.items()
+            for s in study.trials
```

Comment:
A nit, but in Optuna, we use `t` for the trial.
```suggestion
            ("failed_trial", system_attrs["failed_trial"]) in t.system_attrs.items()
            for t in study.trials
```